hammers/query.py

```diff
@@ -284,6 +284,55 @@ def update_orphan_resource_providers(db):
     '''
     return db.query(sql, no_rows=True)
 
+@query
+def get_advance_reservations(db):
+    """Get all advance reservations created by any user"""
+    sql = '''\
+    SELECT lu.name AS user_name, u.extra AS user_extra, l.name AS lease_name, l.id AS lease_id, p.name AS project_name, l.start_date AS start_date 
+    FROM blazar.leases AS l
+    JOIN keystone.user AS u ON l.user_id = u.id 
+    JOIN keystone.local_user AS lu ON u.id = lu.user_id 
+    JOIN keystone.project AS p ON l.project_id = p.id 
+    WHERE l.start_date > UTC_TIMESTAMP()
+    '''
+    
+    return db.query(sql, limit=None)
+
+@query
+def get_idle_leases(db, hours):
+    """
+    Get all leases that have been idle for more than a given number hours.
+    An idle lease is defined as an ongoing lease without active instance and without lease update operation.
+    """
+    sql = '''\
+    SELECT lu.name AS user_name, u.extra AS user_extra, s.lease_name AS lease_name, s.lease_id AS lease_id, p.name AS project_name
+    FROM (
+        SELECT l.id AS lease_id, l.name AS lease_name, l.user_id, l.project_id
+        FROM blazar.leases AS l
+        JOIN blazar.reservations AS r ON l.id = r.lease_id
+        JOIN blazar.computehost_allocations AS ca ON ca.reservation_id = r.id
+        JOIN blazar.computehosts ch on ca.compute_host_id=ch.id
+        LEFT JOIN nova.instances AS i on i.node = ch.hypervisor_hostname
```

Comment:
@priteau @diurnalist ake asked about how to get instances that belong to a lease, and then I released that my previous query was completely wrong. I made changes here and tested it. Could you review the query and see if it makes sense. Also, I fixed all of your comments as well I think. Thanks!

Reply:
Hah, that's a nice coincidence. This one looks OK to me, and if you get good results w/ testing, let's try it out.

Reply:
I have tested it (removing the TIMESTAMPDIFF parts to include a lease that I had just started), and it doesn't seem to work for me. I see several issues:

- the LEFT JOIN is applied on all instance records, we have many of them… you can restrict further in the ON part
- WHERE is applied after LEFT JOIN, so the line doing `i.created_at >= l.start_date` filters leases having had instances, effectively missing leases that have never seen any instance.
- Even with your explanation I still struggle to understand what the part between after OR is actually doing, as it is quite complex. Why are you doing a SUM on ISNULL(i.deleted_at)? Shouldn't it be a COUNT?